include/zephyr/mgmt/osdp.h

```diff
@@ -202,38 +215,251 @@ struct osdp_cmd {
+ * invoked when the PD receives a command from the CP.
  *
- * @retval 0 on success.
- * @retval -1 on failure.
+ * @param cb The callback function's pointer
+ * @param arg A pointer that will be passed as the first argument of `cb`
  */
-int osdp_pd_get_cmd(struct osdp_cmd *cmd);
+void osdp_pd_set_command_callback(pd_commnand_callback_t cb, void *arg);
+
+/**
+ * @brief API to notify PD events to CP. These events are sent to the CP as an
+ * alternate response to a POLL command.
+ *
+ * @param event pointer to event struct. Must be filled by application.
+ *
+ * @retval 0 on success
+ * @retval -1 on failure
+ */
+int osdp_pd_notify_event(struct osdp_event *event);
 
 #else /* CONFIG_OSDP_MODE_PD */
 
-int osdp_cp_set_callback_key_press(
-	int (*cb)(int address, uint8_t key));
-int osdp_cp_set_callback_card_read(
-	int (*cb)(int address, int format, uint8_t *data, int len));
+/* ------------------------------- */
+/*            CP Methods           */
+/* ------------------------------- */
+
+/**
+ * @brief Generic command enqueue API.
+ *
+ * @param pd PD offset number as in `pd_info_t *`.
+ * @param cmd command pointer. Must be filled by application.
+ *
+ * @retval 0 on success
+ * @retval -1 on failure
+ *
+ * @note This method only adds the command on to a particular PD's command
+ * queue. The command itself can fail due to various reasons.
+ */
 int osdp_cp_send_command(int pd, struct osdp_cmd *cmd);
 
+
+/**
+ * @brief Set callback method for CP event notification. This callback is
+ * invoked when the CP receives an event from the PD.
+ *
+ * @param cb The callback function's pointer
+ * @param arg A pointer that will be passed as the first argument of `cb`
+ */
+void osdp_cp_set_event_callback(cp_event_callback_t cb, void *arg);
+
 #endif /* CONFIG_OSDP_MODE_PD */
 
 #ifdef CONFIG_OSDP_SC_ENABLED
 
-uint32_t osdp_get_sc_status_mask(void);
+/* ------------------------------- */
+/*          Common Methods         */
+/* ------------------------------- */
+
+/**
+ * @brief Get a bit mask of number of PD that are online currently.
+ *
+ * @param bitmask pointer to an array of bytes. must be as large as
+ *                (num_pds + 7 / 8).
+ */
+void osdp_get_status_mask(uint8_t *bitmask);
+
+/**
+ * @brief Get a bit mask of number of PD that are online and have an active
+ * secure channel currently.
+ *
+ * @param bitmask pointer to an array of bytes. must be as large as
+ *                (num_pds + 7 / 8).
+ */
+void osdp_get_sc_status_mask(uint8_t *bitmask);
+
+/**
+ * @brief Set osdp_command_complete_callback_t to subscribe to osdp command or
+ * event completion events. This can be used to perform post command actions
+ * such as changing the baud rate of the underlying channel after a COMSET
+ * command was acknowledged/issued by a peer.
+ *
+ * @param ctx OSDP context
+ * @param cb Callback to be invoked when a command is completed.
+ */
+void osdp_set_command_complete_callback(osdp_command_complete_callback_t cb);
```

Comment:
I guess this one should be outside of #ifdef CONFIG_OSDP_SC_ENABLED ... #endIf

Reply:
Thanks for the review, I disabled CONFIG_OSDP_SC_ENABLED and fixed this and other issues that were caused due to code leaks our side this switch.

Reply:
Also, CONFIG_OSDP_SC_ENABLED situation has worsened now more than ever. So I'll add another patch to reduce the number of  ifdefs we use to disable secure channel.